### PR TITLE
UI test improvements

### PR DESCRIPTION
**Pull request by wjordan**
Replaces timing-dependent waits (`'And I press "x" using jQuery'`; `"And I wait for 5 seconds"`) with explicit waits for pages to load or ajax-requests to finish (`'And I press "x" using jQuery to load a new page'`).

This will:

- Reduce test-flakiness due to failures like:
`expected "https://test-studio.code.org/s/allthethings/stage/23/puzzle/2/page/1?noautoplay=true" to include "/page/2" (RSpec::Expectations::ExpectationNotMetError)`, in cases where the timing-dependent wait isn't long enough to fully load the page (due to load-based delay, etc)
- Improve test performance in cases where less time is needed to load the page

### CODE REVIEW

File: dashboard/test/ui/step_definitions/steps.rb [hunk 1]
```diff
@@ -158,6 +179,10 @@ def jquery_is_element_visible(selector)
   wait_with_timeout.until { @browser.execute_script("return $(#{selector.dump}, $(#{parent_selector.dump}).contents()).is(':visible')") }
 end
 
+When /^I wait until Ajax requests are finished$/ do
+  wait_with_short_timeout.until { @browser.execute_script("return $.active == 0") }
```
Comment (joshlory) [inline dashboard/test/ui/step_definitions/steps.rb]: Depends on jQuery?

File: dashboard/test/ui/step_definitions/steps.rb [hunk 2]
```diff
@@ -12,6 +12,27 @@ def wait_with_short_timeout
...
+def element_stale?(element)
+  begin
+    element.enabled?
+    false
+  rescue Selenium::WebDriver::Error::StaleElementReferenceError
+    true
+  end
+end
+
+def page_load(page_load)
```
Comment (joshlory) [inline dashboard/test/ui/step_definitions/steps.rb]: Nit: can we name the argument something more descriptive?  E.g. `should_wait_for_unload`.

File: dashboard/test/ui/step_definitions/steps.rb [hunk 3]
```diff
@@ -698,11 +729,9 @@ def disabled?(selector)
...
 
 Then(/^I reload the page$/) do
-  # Make sure the old page is gone before this step completes, since selenium's navigate.refresh
-  # does not reliably do this for us.
-  @browser.execute_script("if (window) window.seleniumNavigationPending = true;")
-  @browser.navigate.refresh
-  wait_with_short_timeout.until { @browser.execute_script('return !(window && window.seleniumNavigationPending);') }
+  page_load(true) do
+    @browser.navigate.refresh
+  end
```
Comment (islemaster) [inline dashboard/test/ui/step_definitions/steps.rb:732]: :+1: Convenient!

File: dashboard/test/ui/step_definitions/steps.rb (hunk 1 above)
Comment (wjordan) [reply dashboard/test/ui/step_definitions/steps.rb]: Yeah, this approach will only catch requests using `jquery.ajax()`, which currently includes most of the ajax requests in our system but might be different in the future. I can rename this step to something like "Wait until jQuery ajax requests are finished" to make this more clear

**Comment by wjordan** [comment]
~Still seeing flakiness on test build.. Reverting and will work on it some more.~

edit: Actually, the flakiness on the test build happened on the build prior to these changes being merged into the test branch. Confirmed that this PR fixes the flakiness! :tada:

**Comment by breville** [comment]
Very cool!

**Pull request merged**